pkg/router/controller/route_secret_manager.go

```diff
@@ -0,0 +1,217 @@
+func (p *RouteSecretManager) generateSecretHandler(route *routev1.Route) cache.ResourceEventHandlerFuncs {
+	// secret handler
+	secreth := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			secret := obj.(*kapi.Secret)
+			klog.Infof("secret %s added for %s/%s", secret.Name, route.Namespace, route.Name)
+			// Do nothing for add event
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
+			secretOld := old.(*kapi.Secret)
+			secretNew := new.(*kapi.Secret)
+			klog.Infof("secret %s updated: old version %s, new version %s for %s/%s", secretNew.Name, secretOld.ResourceVersion, secretNew.ResourceVersion, route.Namespace, route.Name)
+
+			// re-validate
+			fldPath := field.NewPath("spec").Child("tls").Child("externalCertificate")
+			if err := routeapihelpers.ValidateTLSExternalCertificate(route, fldPath, p.sarClient, p.secretsGetter).ToAggregate(); err != nil {
+				klog.Error(err, "skipping route due to invalid externalCertificate configuration", " route ", route.Name)
+				p.recorder.RecordRouteRejection(route, "ExternalCertificateValidationFailed", err.Error())
+				p.plugin.HandleRoute(watch.Deleted, route)
+			}
+
+			// read referenced secret (updated data)
+			secret, err := p.secretManager.GetSecret(context.TODO(), route.Namespace, route.Name)
+			if err != nil {
+				klog.Error("failed to read referenced secret", err)
+				p.recorder.RecordRouteRejection(route, "ExternalCertificateReadFailed", err.Error())
+				p.plugin.HandleRoute(watch.Deleted, route)
+			}
+
+			// update tls.Certificate and tls.Key
+			route.Spec.TLS.Certificate = string(secret.Data["tls.crt"])
+			route.Spec.TLS.Key = string(secret.Data["tls.key"])
+
+			// call the next plugin with watch.Modified
+			p.plugin.HandleRoute(watch.Modified, route)
+			// commit the changes
+			p.plugin.Commit()
+		},
+		DeleteFunc: func(obj interface{}) {
+			secret := obj.(*kapi.Secret)
+			msg := fmt.Sprintf("secret %s deleted for %s/%s", secret.Name, route.Namespace, route.Name)
+			klog.Info(msg)
+
+			// unregister associated secret monitor
+			if err := p.secretManager.UnregisterRoute(route.Namespace, route.Name); err != nil {
+				klog.Error("failed to unregister route", err)
+			}
+
+			p.recorder.RecordRouteRejection(route, "ExternalCertificateSecretDeleted", msg)
+			p.plugin.HandleRoute(watch.Deleted, route)
+		},
+	}
+	return secreth
+}
+
+func hasExternalCertificate(route *routev1.Route) bool {
+	tls := route.Spec.TLS
+	if tls != nil && tls.ExternalCertificate != nil && len(tls.ExternalCertificate.Name) > 0 {
+		return true
+	}
+	return false
+}
+
+// must be called after hasExternalCertificate
+func getReferencedSecret(route *routev1.Route) string {
+	secretName := route.Spec.TLS.ExternalCertificate.Name
+	klog.Info("Referenced secretName: ", secretName)
```

Comment:
It seems odd that `getReferencedSecret` logs the secret name.  If you keep this log statement, it should have a high log level (I'd say at least 6, maybe 10).

Reply:
I'll keep this for now at the highest log-level (10), incase of any debugging need.